// danger/distribution/runner/DangerfileRunner.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t;
    return { next: verb(0), "throw": verb(1), "return": verb(2) };
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var Runtime = require("jest-runtime");
var jest_config_1 = require("jest-config");
var NodeEnvironment = require("jest-environment-node");
var os = require("os");
var fs = require("fs");
/**
 * Executes a Dangerfile at a specific path, with a context.
 * The values inside a Danger context are applied as globals to the Dangerfiles runtime.
 *
 * @param {DangerContext} dangerfileContext the global danger context
 * @returns {any} the results of the run
 */
function createDangerfileRuntimeEnvironment(dangerfileContext) {
    return __awaiter(this, void 0, void 0, function () {
        var config, environment, runnerGlobal, context, prop, anyContext, hasteConfig, hasteMap, resolver, runtime;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, dangerJestConfig()];
                case 1:
                    config = _a.sent();
                    environment = new NodeEnvironment(config);
                    runnerGlobal = environment.global;
                    context = dangerfileContext;
                    // Adds things like fail, warn ... to global
                    for (prop in context) {
                        if (context.hasOwnProperty(prop)) {
                            anyContext = context;
                            runnerGlobal[prop] = anyContext[prop];
                        }
                    }
                    hasteConfig = { automock: false, maxWorkers: 1, resetCache: false };
                    return [4 /*yield*/, Runtime.createHasteMap(config, hasteConfig).build()];
                case 2:
                    hasteMap = _a.sent();
                    resolver = Runtime.createResolver(config, hasteMap.moduleMap);
                    runtime = new Runtime(config, environment, resolver);
                    return [2 /*return*/, {
                            context: context,
                            environment: environment,
                            runtime: runtime
                        }];
            }
        });
    });
}
exports.createDangerfileRuntimeEnvironment = createDangerfileRuntimeEnvironment;
/**
 * The Jest config object for this Danger run
 * @returns {any} the results of the run
 */
function dangerJestConfig() {
    return __awaiter(this, void 0, void 0, function () {
        var jestConfig;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, jest_config_1.readConfig([], process.cwd())];
                case 1:
                    jestConfig = _a.sent();
                    return [2 /*return*/, {
                            cacheDirectory: os.tmpdir(),
                            setupFiles: [],
                            name: "danger",
                            testEnvironment: "node",
                            haste: {
                                defaultPlatform: "danger-js"
                            },
                            moduleNameMapper: [],
                            moduleDirectories: ["node_modules"],
                            moduleFileExtensions: ["js"].concat(jestConfig.config.moduleFileExtensions),
                            transform: [["js$", "babel-jest"]].concat(jestConfig.config.transform),
                            testPathIgnorePatterns: jestConfig.config.testPathIgnorePatterns,
                            cache: null,
                            testRegex: "",
                            testPathDirs: [process.cwd()],
                            transformIgnorePatterns: ["/node_modules/"]
                        }];
            }
        });
    });
}
exports.dangerJestConfig = dangerJestConfig;
/**
 * Executes a Dangerfile at a specific path, with a context.
 * The values inside a Danger context are applied as globals to the Dangerfiles runtime.
 *
 * @param {string} filename the file path for the dangerfile
 * @param {any} environment the results of createDangerfileRuntimeEnvironment
 * @returns {DangerResults} the results of the run
 */
function runDangerfileEnvironment(filename, environment) {
    return __awaiter(this, void 0, void 0, function () {
        var runtime, results;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    runtime = environment.runtime;
                    // Require our dangerfile
                    ensureCleanDangerfile(filename, function () {
                        runtime.requireModule(filename);
                    });
                    results = environment.context.results;
                    return [4 /*yield*/, Promise.all(results.scheduled.map(function (fnOrPromise) {
                            if (fnOrPromise instanceof Promise) {
                                return fnOrPromise;
                            }
                            if (fnOrPromise.length === 1) {
                                // callback-based function
                                return new Promise(function (res) { return fnOrPromise(res); });
                            }
                            return fnOrPromise();
                        }))];
                case 1:
                    _a.sent();
                    return [2 /*return*/, {
                            fails: results.fails,
                            warnings: results.warnings,
                            messages: results.messages,
                            markdowns: results.markdowns,
                        }];
            }
        });
    });
}
exports.runDangerfileEnvironment = runDangerfileEnvironment;
/**
 * Passes in a dangerfile path, will remove any references to import/require `danger`
 * then runs the internal closure with a "safe" version of the Dangerfile.
 * Then it will clean itself up afterwards, and use the new version.
 *
 * Note: We check for equality to not trigger the jest watcher for tests.
 *
 * @param {string} filename the file path for the dangerfile
 * @param {Function} closure code to run with a cleaned Dangerfile
 * @returns {void}
 */
function ensureCleanDangerfile(filename, closure) {
    var originalContents = fs.readFileSync(filename).toString();
    updateDangerfile(filename);
    closure();
    if (originalContents !== fs.readFileSync(filename).toString()) {
        fs.writeFileSync(filename, originalContents);
    }
}
/**
 * Updates a Dangerfile to remove the import for Danger
 * @param {string} filename the file path for the dangerfile
 * @returns {void}
 */
function updateDangerfile(filename) {
    var contents = fs.readFileSync(filename).toString();
    var cleanedDangerFile = cleanDangerfile(contents);
    if (contents !== cleanedDangerFile) {
        fs.writeFileSync(filename, cleanDangerfile(contents));
    }
}
exports.updateDangerfile = updateDangerfile;
// https://regex101.com/r/dUq4yB/1
var requirePattern = /^.* require\(('|")danger('|")\);?$/gm;
//  https://regex101.com/r/dUq4yB/2
var es6Pattern = /^.* from ('|")danger('|");?$/gm;
/**
 * Updates a Dangerfile to remove the import for Danger
 * @param {string} contents the file path for the dangerfile
 * @returns {string} the revised Dangerfile
 */
function cleanDangerfile(contents) {
    return contents
        .replace(es6Pattern, "// Removed import")
        .replace(requirePattern, "// Removed require");
}
exports.cleanDangerfile = cleanDangerfile;
